auraed/src/vms/virtual_machine.rs

```diff
@@ -0,0 +1,332 @@
+            watchdog: false,
+            pci_segments: None,
+            platform: None,
+            tpm: None,
+            preserved_fds: None,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct NetSpec {
+    pub tap: Option<String>,
+    pub ip: Ipv4Addr,
+    pub mask: Ipv4Addr,
+    pub mac: MacAddr,
+    pub host_mac: Option<MacAddr>,
+}
+
+impl From<NetSpec> for vmm::vm_config::NetConfig {
+    fn from(spec: NetSpec) -> Self {
+        vmm::vm_config::NetConfig {
+            tap: spec.tap,
+            ip: spec.ip,
+            mask: spec.mask,
+            mac: spec.mac,
+            host_mac: spec.host_mac,
+            mtu: None,
+            iommu: false,
+            num_queues: DEFAULT_NET_NUM_QUEUES,
+            queue_size: DEFAULT_NET_QUEUE_SIZE,
+            vhost_user: false,
+            vhost_socket: None,
+            vhost_mode: VhostMode::default(),
+            id: None,
+            fds: None,
+            rate_limiter_config: None,
+            pci_segment: 0,
+            offload_tso: false,
+            offload_ufo: false,
+            offload_csum: false,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct MountSpec {
+    pub host_path: PathBuf,
+    pub read_only: bool,
+}
+
+impl From<MountSpec> for vmm::vm_config::DiskConfig {
+    fn from(spec: MountSpec) -> Self {
+        vmm::vm_config::DiskConfig {
+            path: Some(spec.host_path),
+            readonly: spec.read_only,
+            direct: false,
+            iommu: false,
+            num_queues: DEFAULT_DISK_NUM_QUEUES,
+            queue_size: DEFAULT_DISK_QUEUE_SIZE,
+            vhost_user: false,
+            vhost_socket: None,
+            rate_limit_group: None,
+            rate_limiter_config: None,
+            id: None,
+            disable_io_uring: false,
+            disable_aio: false,
+            pci_segment: 0,
+            serial: None,
+            queue_affinity: None,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct VirtualMachine {
+    pub id: VmID,
+    pub vm: VmSpec,
+    status: VmStatus,
+    manager: Arc<Mutex<Manager>>,
+}
+
+#[derive(Debug, Clone)]
+enum VmStatus {
+    Running,
+    Stopped,
+    Deleting,
+}
+
+impl fmt::Debug for VirtualMachine {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "VirtualMachine {{ id: {:?}, vm: {:?} }}", self.id, self.vm)
+    }
+}
+
+impl VirtualMachine {
+    pub fn new(id: VmID, spec: VmSpec) -> Result<Self, anyhow::Error> {
+        let mut manager = Manager::new();
+        manager.start()?;
+
+        if let Some(sender) = &manager.sender {
+            vmm::api::VmCreate
+                .send(
+                    manager.events.try_clone()?,
+                    sender.clone(),
+                    Arc::new(Mutex::new(spec.clone().into())),
+                )
+                .expect("Failed to send create request");
+        } else {
+            return Err(anyhow!("Virtual machine manager not initialized"));
```

Comment:
Should add custom error types for the VM module.

Reply:
yes please! though a TODO is fine for now and someone can pick it up as a quick win.